### DISCUSSION TITLE
Ramping laser power with Fusions CO2 (in extraction script)

### DISCUSSION THREAD

**Discussion by cmmercer** [NMGRL/pychron#2191] [Q&A] [answered]
I'm trying to get our Fusions CO2 laser ramping (previously our extraction script was just setting it to the target power), but I'm running into an issue where the ramp starts and gets hung on the first step of the ramp without moving on to subsequent steps. This seems like the `while` loop in the Ramper is not iterating (like it hits a false condition immediately and exits)?

Here is a snippet with the relevant portion of the extraction script:
```
def do_extraction():
    # laser is enabled before this function is called
    if ramp_duration > 0:
        ramp_rate = extract_value/ramp_duration # This was zero from the experiment context, so compute manually
        if (extract_units == 'watts'):
            info('Ramping laser power to {} {} at {} {}/s'.format(extract_value, extract_units, ramp_rate, extract_units))
            begin_interval(ramp_duration + duration)
            # I tried just setting duration (without rate), but got similar behavior
            ramp(setpoint=extract_value, duration=ramp_duration, rate=ramp_rate, period=0.5)
        else:
            warning('Unsupported extraction units ({}) in argus_unknown.py extraction script'.format(extract_units))
            warning('Canceling run')
            cancel()
            return
    else:
        info('Setting laser power to {} {}'.format(extract_value, extract_units))
        extract(extract_value, extract_units)
    complete_interval()
    info('Laser heating complete') # laser is disabled when this function returns to the call point
```

All I see in the console is this: 
```
11:01:01 -- Ramping laser power to 10.0 watts at 0.5 watts/s
11:01:01 -- BEGIN INTERVAL Interval 1 waiting for 80.0
11:01:01 -- ramp step 1. setpoint=0.25
11:01:01 -- COMPLETE INTERVAL waiting for Interval 1 to complete
```

For now, I'm going to work around it by building my own loop and calling the standard `extract(args)` function.

**Comment by jirhiker** [comment, marked as answer]
can you post a log file for this. That will give more information as to why the ramp is exiting

**Reply by jirhiker** [reply]
I also notice you said "measurement script"

Interaction with the laser should happen in the extraction script

**Reply by cmmercer** [reply]
Whoops, I meant the extraction script; updated discussion title.
Here is a log of my last attempt that used the `Ramper.ramp(...)` function. I'm not sure if it's due to a thread block with `begin_interval`, or what could be causing the while loop to exit...
[3101-32-004.log](https://github.com/NMGRL/pychron/files/10403730/3101-32-004.log)

**Reply by jirhiker** [reply]
Thanks

Looks like a bug in `LaserManager.set_laser_power`.  Ramp is expecting `set_laser_power` to return True in order to continue. 

Fixed `set_laser_power` so that it returns `True` if the power is successfully set

**Reply by jirhiker** [reply]
See https://github.com/NMGRL/pychron/commit/bfaab36928570f7c68948154bb9ff65f502507cc
